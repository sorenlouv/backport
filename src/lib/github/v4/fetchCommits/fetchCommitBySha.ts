import { graphql } from '../../../../graphql/generated';
import { ValidConfigOptions } from '../../../../options/options';
import { BackportError } from '../../../BackportError';
import { isMissingConfigFileException } from '../../../remoteConfig';
import {
  Commit,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { GithubV4Exception, getGraphQLClient } from '../client/graphqlClient';

export async function fetchCommitBySha(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  repoName: string;
  repoOwner: string;
  sha: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    repoName,
    repoOwner,
    sha,
    sourceBranch,
  } = options;

  const query = graphql(`
    query CommitsBySha($repoOwner: String!, $repoName: String!, $sha: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $sha) {
          __typename
          ...SourceCommitWithTargetPullRequestFragment
        }
      }
    }
  `);

  const variables = { repoOwner, repoName, sha };
  const client = getGraphQLClient({ accessToken, githubApiBaseUrlV4 });
  const result = await client.query(query, variables);

  if (result.error && !isMissingConfigFileException(result)) {
    throw new GithubV4Exception(result);
  }

  const { data } = result;

  const sourceCommit = data?.repository?.object;
  if (sourceCommit?.__typename !== 'Commit') {
    throw new BackportError(
      `No commit found on branch "${sourceBranch}" with sha "${sha}"`,
    );
  }

  return parseSourceCommit({ options, sourceCommit });
}
