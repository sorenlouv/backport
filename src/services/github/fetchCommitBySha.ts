import axios from 'axios';
import isEmpty from 'lodash.isempty';
import { BackportOptions } from '../../options/options';
import { HandledError } from '../HandledError';
import { CommitSelected } from './Commit';
import { GithubCommit, GithubSearch } from './GithubApiTypes';
import { handleGithubError } from './handleGithubError';
import { getFormattedCommitMessage } from './commitFormatters';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const { apiHostname, repoName, repoOwner, sha, accessToken } = options;
  try {
    const res = await axios.get<GithubSearch<GithubCommit>>(
      `https://${apiHostname}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found on master with sha "${sha}"`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;

    const message = getFormattedCommitMessage({
      message: commitRes.commit.message,
      sha: fullSha
    });

    return {
      message,
      sha: fullSha
    };
  } catch (e) {
    throw handleGithubError(e);
  }
}
