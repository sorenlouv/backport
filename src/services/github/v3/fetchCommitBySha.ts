import isEmpty from 'lodash.isempty';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { CommitSelected } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import { getFormattedCommitMessage, getShortSha } from '../commitFormatters';
import { apiRequestV3 } from './apiRequestV3';

export interface GithubCommit {
  commit: {
    message: string;
  };
  sha: string;
}

interface GithubSearch<T> {
  items: T[];
}

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const {
    githubApiBaseUrlV3,
    repoName,
    repoOwner,
    sha,
    accessToken,
    username,
  } = options;

  const spinner = ora(`Loading commit "${getShortSha(sha)}"`).start();

  try {
    const res = await apiRequestV3<GithubSearch<GithubCommit>>({
      method: 'get',
      url: `${githubApiBaseUrlV3}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1`,
      auth: {
        username: username,
        password: accessToken,
      },
      headers: {
        Accept: 'application/vnd.github.cloak-preview',
      },
    });

    spinner.stop();

    // TODO: it should be possible to backport from other branches than master
    if (isEmpty(res.items)) {
      throw new HandledError(`No commit found on master with sha "${sha}"`);
    }

    const commitRes = res.items[0];
    const fullSha = commitRes.sha;

    const formattedMessage = getFormattedCommitMessage({
      message: commitRes.commit.message,
      sha: fullSha,
    });

    return {
      sourceBranch: 'master',
      selectedTargetBranches: [],
      formattedMessage,
      sha: fullSha,
    };
  } catch (e) {
    spinner.fail();
    throw e;
  }
}
