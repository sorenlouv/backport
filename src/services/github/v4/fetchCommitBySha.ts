import chalk from 'chalk';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { Commit } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import {
  getFormattedCommitMessage,
  getShortSha,
  getPullNumberFromMessage,
} from '../commitFormatters';
import { apiRequestV4 } from './apiRequestV4';
import {
  PullRequestNode,
  pullRequestFragment,
  pullRequestFragmentName,
  getExistingTargetPullRequests,
  getPullRequestLabels,
} from './getExistingTargetPullRequests';
import { getTargetBranchesFromLabels } from './getTargetBranchesFromLabels';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<Commit> {
  const { accessToken, githubApiBaseUrlV4, repoName, repoOwner } = options;

  const query = /* GraphQL */ `
  query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
    repository(owner: $repoOwner, name: $repoName) {
      object(expression: $oid) {
        ... on Commit {
          message
          oid
          associatedPullRequests(first: 1) {
            edges {
              node {
                ...${pullRequestFragmentName}
              }
            }
          }
        }
      }
    }
  }

    ${pullRequestFragment}
  `;

  const spinner = ora(`Loading commit "${getShortSha(options.sha)}"`).start();

  let res: DataResponse;
  try {
    res = await apiRequestV4<DataResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        oid: options.sha,
      },
    });
    spinner.stop();
  } catch (e) {
    spinner.fail();
    throw e;
  }

  if (!res.repository.object) {
    throw new HandledError(
      `No commit found on branch "${options.sourceBranch}" with sha "${options.sha}"`
    );
  }

  const sha = res.repository.object.oid;
  const commitMessage = res.repository.object.message;
  const pullRequestNode =
    res.repository.object.associatedPullRequests.edges?.[0]?.node;

  const pullNumber =
    pullRequestNode?.number || getPullNumberFromMessage(commitMessage);

  const sourceBranch = pullRequestNode?.baseRefName || options.sourceBranch;

  const formattedMessage = getFormattedCommitMessage({
    message: commitMessage,
    pullNumber,
    sha,
  });

  spinner.stopAndPersist({
    symbol: chalk.green('?'),
    text: `${chalk.bold('Select commit')} ${chalk.cyan(formattedMessage)}`,
  });

  const existingTargetPullRequests = getExistingTargetPullRequests(
    commitMessage,
    pullRequestNode
  );

  const targetBranchesFromLabels = getTargetBranchesFromLabels({
    existingTargetPullRequests,
    branchLabelMapping: options.branchLabelMapping,
    labels: getPullRequestLabels(pullRequestNode),
  });

  return {
    sourceBranch,
    targetBranchesFromLabels,
    sha,
    formattedMessage,
    originalMessage: commitMessage,
    pullNumber,
    existingTargetPullRequests,
  };
}

interface DataResponse {
  repository: {
    object: {
      message: string;
      oid: string;
      associatedPullRequests: {
        edges:
          | {
              node: PullRequestNode;
            }[]
          | null;
      };
    } | null;
  };
}
