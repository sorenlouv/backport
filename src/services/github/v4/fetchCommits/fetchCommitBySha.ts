import { ValidConfigOptions } from '../../../../options/options';
import { HandledError } from '../../../HandledError';
import {
  Commit,
  SourceCommitWithTargetPullRequest,
  sourceCommitWithTargetPullRequestFragment,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { apiRequestV4 } from '../apiRequestV4';

export async function fetchCommitBySha(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  historicalBranchLabelMappings: ValidConfigOptions['historicalBranchLabelMappings'];
  repoName: string;
  repoOwner: string;
  sha: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    repoName,
    repoOwner,
    sha,
    sourceBranch,
  } = options;

  const query = /* GraphQL */ `
    query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $oid) {
          ...SourceCommitWithTargetPullRequest
        }
      }
    }

    ${sourceCommitWithTargetPullRequestFragment.source}
  `;

  const res = await apiRequestV4<CommitsByShaResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: {
      repoOwner,
      repoName,
      oid: sha,
    },
  });

  const sourceCommit = res.repository.object;
  if (!sourceCommit) {
    throw new HandledError(
      `No commit found on branch "${sourceBranch}" with sha "${sha}"`
    );
  }

  return parseSourceCommit({ options, sourceCommit });
}

interface CommitsByShaResponse {
  repository: {
    object: SourceCommitWithTargetPullRequest | null;
  };
}
