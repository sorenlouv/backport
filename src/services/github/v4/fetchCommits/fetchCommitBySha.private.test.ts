import { getDevAccessToken } from '../../../../test/private/getDevAccessToken';
import { fetchCommitBySha } from './fetchCommitBySha';

describe('fetchCommitBySha', () => {
  let devAccessToken: string;

  beforeEach(async () => {
    devAccessToken = await getDevAccessToken();
  });

  it('should return single commit with pull request', async () => {
    await expect(
      await fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken: devAccessToken,
        sha: 'cb6fbc0',
        sourceBranch: 'master',
        historicalBranchLabelMappings: [],
      })
    ).toEqual({
      committedDate: '2020-07-07T20:40:28Z',
      originalMessage: '[APM] Add API tests (#70740)',
      pullNumber: 70740,
      pullUrl: 'https://github.com/elastic/kibana/pull/70740',
      sha: 'cb6fbc0e1b406675724181a3e9f59459b5f8f892',
      sourceBranch: 'master',
      expectedTargetPullRequests: [
        {
          branch: '7.x',
          state: 'MERGED',
          number: 71014,
          url: 'https://github.com/elastic/kibana/pull/71014',
        },
      ],
    });
  });

  it('throws if sha does not exist', async () => {
    await expect(
      fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken: devAccessToken,
        sha: 'fc22f59',
        sourceBranch: 'main',
        historicalBranchLabelMappings: [],
      })
    ).rejects.toThrowError(
      'No commit found on branch "main" with sha "fc22f59"'
    );
  });

  it('throws if sha is invalid', async () => {
    await expect(
      fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken: devAccessToken,
        sha: 'myCommitSha',
        sourceBranch: 'main',
        historicalBranchLabelMappings: [],
      })
    ).rejects.toThrowError(
      'No commit found on branch "main" with sha "myCommitSha"'
    );
  });
});
