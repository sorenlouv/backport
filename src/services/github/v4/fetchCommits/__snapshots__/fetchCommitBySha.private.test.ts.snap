// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitBySha snapshot request/response makes the right queries: Query: CommitsBySha 1`] = `
"
    query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $oid) {
          ...SourceCommitWithTargetPullRequest
        }
      }
    }

    
    fragment SourceCommitWithTargetPullRequest on Commit {
      # Source Commit
      repository {
        name
        owner {
          login
        }
      }
      oid
      message
      committedDate

      # Source pull request: PR where source commit was merged in
      associatedPullRequests(first: 1) {
        edges {
          node {
            url
            number
            labels(first: 50) {
              nodes {
                name
              }
            }
            baseRefName

            # source merge commit (the commit that actually went into the source branch)
            sourceMergeCommit: mergeCommit {
              oid
              message
            }

            # (possible) backport pull requests referenced in the source pull request
            timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
              edges {
                node {
                  ... on CrossReferencedEvent {
                    targetPullRequest: source {
                      __typename

                      # Target PRs (backport PRs)
                      ... on PullRequest {
                        # target merge commit: the backport commit that was merged into the target branch
                        targetMergeCommit: mergeCommit {
                          oid
                          message
                        }
                        repository {
                          name
                          owner {
                            login
                          }
                        }
                        url
                        title
                        state
                        baseRefName
                        number
                        commits(first: 20) {
                          edges {
                            node {
                              targetCommit: commit {
                                message
                                oid
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  
  "
`;

exports[`fetchCommitBySha snapshot request/response returns the correct response 1`] = `
Object {
  "committedDate": "2021-12-20T14:20:16Z",
  "expectedTargetPullRequests": Array [
    Object {
      "branch": "8.0",
      "mergeCommit": Object {
        "message": "[APM] Add note about synthtrace to APM docs (#121633) (#121643)

Co-authored-by: SÃ¸ren Louv-Jansen <soren.louv@elastic.co>",
        "sha": "842adfdeb5541b059231857522f9009771a46107",
      },
      "number": 121643,
      "state": "MERGED",
      "url": "https://github.com/elastic/kibana/pull/121643",
    },
  ],
  "originalMessage": "[APM] Add note about synthtrace to APM docs (#121633)",
  "pullNumber": 121633,
  "pullUrl": "https://github.com/elastic/kibana/pull/121633",
  "sha": "d421ddcf6157150596581c7885afa3690cec6339",
  "sourceBranch": "main",
}
`;
